pycarol/sql.py

```diff
@@ -64,38 +82,52 @@ def query(self, query, params=None, method='sync', dataframe=True,  **kw):
                 if socket runs the query using web socket and returns the result
             dataframe: `bool` default `True`
                 if True returns a pandas dataframe
+            service_account: in case you have a service account for accessing BigQuery.
             **kw: `dict`
         Returns:
-        
+
         """
-        if method not in ('sync', 'socket'):
-            raise ValueError("kind must be either 'sync' or 'socket'")
-        
+        methods = ("sync", "socket", "bigquery")
+
         params = params or []
-        payload = {"preparedStatement": query, "params": params} 
-        
-        if method == 'socket':
-            last_result = False
-            results = []
-            with SQLSocket(self.carol, **kw) as ws:
-                resp = self.carol.call_api(path='v2/sql/query', method='POST', data=payload)
-                while not last_result:
-                    resp = ws.recv()
-                    resp = json.loads(resp)
-                    result = json.loads(resp['message'])
-                    last_result = result['lastResult']
-                    if result['success']:
-                        results.extend(result['results'])
-                    else:
-                        raise ValueError(f"{result['error']}")
-            
-        elif method == 'sync':
-            results = self.carol.call_api(path='v2/sql/querySync', method='POST', data=payload)
-            if not results[0]['success']:
-                raise ValueError(f"{results[0]['error']}")
-            results = list(itertools.chain(*[i['results'] for i in results]))
+        payload = {"preparedStatement": query, "params": params}
+        if method == "socket":
+            results = _socket_query(self.carol, payload, **kw)
+        elif method == "sync":
+            results = _sync_query(self.carol, payload)
+        elif method == "bigquery":
+            results = bigquery.query(self.carol, query, service_account)
+        else:
+            raise ValueError(f"'method' must be either: {methods}")
+
         if dataframe:
-            import pandas as pd
             return pd.DataFrame(results)
-        else:
-            return results
\ No newline at end of file
+        if not dataframe and method == "bigquery":
+            records = [dict(row) for row in results]
+            return json.dumps(str(records))
```

Comment:
why dump to string and not return the dict?

Reply:
changed to dict.